drake/systems/LinearSystem.h

```diff
@@ -1,38 +1,64 @@
 #ifndef DRAKE_LINEARSYSTEM_H
 #define DRAKE_LINEARSYSTEM_H
 
-#include "DrakeSystem.h"
+#include "System.h"
+
+namespace Drake {
 
 /// Implements
 ///   xcdot = Ac*x + Bc*u + xcdot0
 ///   xdn = Ad*x + Bd*u + xdn0
 ///   y = C*x + D*u + y0
 
-class AffineSystem : public DrakeSystem {
+template <template<typename> class StateVector, template<typename> class InputVector, template<typename> class OutputVector, bool _isDirectFeedthrough = true >
+class AffineSystem : public System< AffineSystem<StateVector,InputVector,OutputVector,_isDirectFeedthrough>, StateVector, InputVector, OutputVector, false, _isDirectFeedthrough> {
 public:
-  AffineSystem(const std::string& name,
-               const Eigen::MatrixXd& _Ac,const Eigen::MatrixXd& _Bc,const Eigen::VectorXd& _xcdot0,
-               const Eigen::MatrixXd& _Ad,const Eigen::MatrixXd& _Bd,const Eigen::VectorXd& _xdn0,
-               const Eigen::MatrixXd& _C,const Eigen::MatrixXd& _D,const Eigen::VectorXd& _y0);
+  typedef System< AffineSystem<StateVector,InputVector,OutputVector,_isDirectFeedthrough>, StateVector, InputVector, OutputVector, false, _isDirectFeedthrough> BaseType;
+  using BaseType::num_states;
+  using BaseType::num_inputs;
+  using BaseType::num_outputs;
+
+  AffineSystem(const Eigen::Matrix<double,num_states,num_states>& A,const Eigen::Matrix<double,num_states,num_inputs>& B,const Eigen::Matrix<double,num_states,1>& xdot0,
+               const Eigen::Matrix<double,num_outputs,num_states>& C,const Eigen::Matrix<double,num_outputs,num_inputs>& D,const Eigen::Matrix<double,num_outputs,1>& y0)
+          : A(A),B(B),C(C),D(D),xdot0(xdot0),y0(y0) {
+    assert(_isDirectFeedthrough || D.isZero() );
+  }
 
-  virtual Eigen::VectorXd dynamics(double t, const Eigen::VectorXd& x, const Eigen::VectorXd& u) const override {
-    return Ac*x + Bc*u + xcdot0;
+  template <typename ScalarType>
+  StateVector<ScalarType> dynamicsImplementation(const StateVector<ScalarType>& x, const InputVector<ScalarType>& u) const {
+    using namespace Eigen;
+    StateVector<ScalarType> xdot = A*static_cast<Matrix<ScalarType,num_states,1> >(x) + B*static_cast<Matrix<ScalarType,num_inputs,1> >(u) + xdot0;
```

Comment:
This kind of usage makes me think that it might be better to just require that for any state vector there exists a toMatrix function that returns an eigen matrix of the appropriate size.

For later.